From 33acc2ecf56d4c103adb4c1e55480dbdd3fda4dd Mon Sep 17 00:00:00 2001
From: Minggui Cao <minggui.cao@intel.com>
Date: Tue, 30 Mar 2021 11:08:48 +0800
Subject: [PATCH] rebase CAT patch to release-v2.4

1. based on commit: 9e4fd1f1193353c5ef
2. this patch just supports CAT:
     boards: TGL/EHL/WHL-i7
     scenarios: hybrid_rt/industry

Signed-off-by: Minggui Cao <minggui.cao@intel.com>
---
 hypervisor/arch/x86/guest/vmsr.c              | 10 ++---
 hypervisor/arch/x86/rdt.c                     | 38 ++++++++++++++++++-
 .../config_tools/data/ehl-crb-b/ehl-crb-b.xml |  6 +--
 .../config_tools/data/ehl-crb-b/hybrid_rt.xml | 25 ++++--------
 misc/config_tools/data/ehl-crb-b/industry.xml | 24 +++---------
 misc/config_tools/data/tgl-rvp/hybrid_rt.xml  |  8 +++-
 misc/config_tools/data/tgl-rvp/industry.xml   |  8 +++-
 misc/config_tools/data/tgl-rvp/tgl-rvp.xml    |  6 +--
 .../data/whl-ipc-i7/hybrid_rt.xml             |  8 +++-
 .../config_tools/data/whl-ipc-i7/industry.xml |  8 +++-
 .../data/whl-ipc-i7/whl-ipc-i7.xml            |  3 ++
 11 files changed, 86 insertions(+), 58 deletions(-)

diff --git a/hypervisor/arch/x86/guest/vmsr.c b/hypervisor/arch/x86/guest/vmsr.c
index bae2bff2c..4cdbff6ab 100644
--- a/hypervisor/arch/x86/guest/vmsr.c
+++ b/hypervisor/arch/x86/guest/vmsr.c
@@ -327,13 +327,9 @@ static void init_msr_area(struct acrn_vcpu *vcpu)
 
 	/* only load/restore MSR IA32_PQR_ASSOC when hv and guest have differnt settings */
 	if (is_platform_rdt_capable() && (vcpu_clos != hv_clos)) {
-		vcpu->arch.msr_area.guest[MSR_AREA_IA32_PQR_ASSOC].msr_index = MSR_IA32_PQR_ASSOC;
-		vcpu->arch.msr_area.guest[MSR_AREA_IA32_PQR_ASSOC].value = clos2pqr_msr(vcpu_clos);
-		vcpu->arch.msr_area.host[MSR_AREA_IA32_PQR_ASSOC].msr_index = MSR_IA32_PQR_ASSOC;
-		vcpu->arch.msr_area.host[MSR_AREA_IA32_PQR_ASSOC].value = clos2pqr_msr(hv_clos);
-		vcpu->arch.msr_area.count++;
-		pr_acrnlog("switch clos for VM %u vcpu_id %u, host 0x%x, guest 0x%x",
-			vcpu->vm->vm_id, vcpu->vcpu_id, hv_clos, vcpu_clos);
+		msr_write_pcpu(MSR_IA32_PQR_ASSOC, clos2pqr_msr(vcpu_clos), pcpuid_from_vcpu(vcpu));
+		pr_acrnlog("switch clos for VM %u vcpu_id %u: 0x%x",
+			vcpu->vm->vm_id, vcpu->vcpu_id, vcpu_clos);
 	}
 }
 
diff --git a/hypervisor/arch/x86/rdt.c b/hypervisor/arch/x86/rdt.c
index e33e1b3aa..ef7f0f486 100644
--- a/hypervisor/arch/x86/rdt.c
+++ b/hypervisor/arch/x86/rdt.c
@@ -60,6 +60,34 @@ static struct rdt_info res_cap_info[RDT_NUM_RESOURCES] = {
 	},
 };
 
+static void check_init_L2_cat(void)
+{
+	struct cpuinfo_x86 *cpu_info = get_pcpu_info();
+
+	if ((cpu_info->displayfamily == 6) &&
+		(cpu_info->displaymodel == 0x96 /* EHL */)) {
+
+		res_cap_info[RDT_RESOURCE_L2].clos_max = MAX_CACHE_CLOS_NUM_ENTRIES;
+
+		platform_l2_clos_array[0].value.clos_mask = 0x0ff;
+		platform_l2_clos_array[1].value.clos_mask = 0xf00;
+		platform_l2_clos_array[2].value.clos_mask = 0x00f;
+		platform_l2_clos_array[3].value.clos_mask = 0x001;
+	}
+}
+
+static inline bool is_follow_spec()
+{
+	struct cpuinfo_x86 *cpu_info = get_pcpu_info();
+
+	if (cpu_info->displayfamily == 6 &&
+		((cpu_info->displaymodel == 0x8c /* TGL-U */) ||
+		(cpu_info->displaymodel == 0x96 /* EHL */))) {
+		return true;
+	}
+	return false;
+}
+
 /*
  * @pre res == RDT_RESOURCE_L3 || res == RDT_RESOURCE_L2
  */
@@ -114,7 +142,12 @@ void init_rdt_info(void)
 	uint8_t i;
 	uint32_t eax = 0U, ebx = 0U, ecx = 0U, edx = 0U;
 
-	if (pcpu_has_cap(X86_FEATURE_RDT_A)) {
+	if (MAX_CACHE_CLOS_NUM_ENTRIES > 0) {
+		printf("LLC CAT enabled, %sfollow spec\n", is_follow_spec() ? "" : "not ");
+		res_cap_info[RDT_RESOURCE_L3].clos_max = MAX_CACHE_CLOS_NUM_ENTRIES;
+
+		check_init_L2_cat();
+	} else if (pcpu_has_cap(X86_FEATURE_RDT_A)) {
 		cpuid_subleaf(CPUID_RDT_ALLOCATION, 0U, &eax, &ebx, &ecx, &edx);
 
 		/* If HW supports L3 CAT, EBX[1] is set */
@@ -197,6 +230,9 @@ uint64_t clos2pqr_msr(uint16_t clos)
 {
 	uint64_t pqr_assoc;
 
+	if (!is_follow_spec()) {
+		return clos;
+	}
 	pqr_assoc = msr_read(MSR_IA32_PQR_ASSOC);
 	pqr_assoc = (pqr_assoc & 0xffffffffUL) | ((uint64_t)clos << 32U);
 
diff --git a/misc/config_tools/data/ehl-crb-b/ehl-crb-b.xml b/misc/config_tools/data/ehl-crb-b/ehl-crb-b.xml
index 300b37791..1aef907e8 100644
--- a/misc/config_tools/data/ehl-crb-b/ehl-crb-b.xml
+++ b/misc/config_tools/data/ehl-crb-b/ehl-crb-b.xml
@@ -220,9 +220,9 @@
 	</TPM_INFO>
 
 	<CLOS_INFO>
-	rdt resources supported: L2
-	rdt resource clos max: 16
-	rdt resource mask max: '0xfff'
+	rdt resources supported: L3
+	rdt resource clos max: 4
+	rdt resource mask max: '0xffff'
 	</CLOS_INFO>
 
 	<IOMEM_INFO>
diff --git a/misc/config_tools/data/ehl-crb-b/hybrid_rt.xml b/misc/config_tools/data/ehl-crb-b/hybrid_rt.xml
index fc5112cd7..213ea1e68 100644
--- a/misc/config_tools/data/ehl-crb-b/hybrid_rt.xml
+++ b/misc/config_tools/data/ehl-crb-b/hybrid_rt.xml
@@ -16,24 +16,13 @@
             <MULTIBOOT2>y</MULTIBOOT2>
             <ENFORCE_TURNOFF_AC>y</ENFORCE_TURNOFF_AC>
             <RDT>
-                <RDT_ENABLED>n</RDT_ENABLED>
+                <RDT_ENABLED>y</RDT_ENABLED>
                 <CDP_ENABLED>n</CDP_ENABLED>
-                <CLOS_MASK>0xfff</CLOS_MASK>
-                <CLOS_MASK>0xfff</CLOS_MASK>
-                <CLOS_MASK>0xfff</CLOS_MASK>
-                <CLOS_MASK>0xfff</CLOS_MASK>
-                <CLOS_MASK>0xfff</CLOS_MASK>
-                <CLOS_MASK>0xfff</CLOS_MASK>
-                <CLOS_MASK>0xfff</CLOS_MASK>
-                <CLOS_MASK>0xfff</CLOS_MASK>
-                <CLOS_MASK>0xfff</CLOS_MASK>
-                <CLOS_MASK>0xfff</CLOS_MASK>
-                <CLOS_MASK>0xfff</CLOS_MASK>
-                <CLOS_MASK>0xfff</CLOS_MASK>
-                <CLOS_MASK>0xfff</CLOS_MASK>
-                <CLOS_MASK>0xfff</CLOS_MASK>
-                <CLOS_MASK>0xfff</CLOS_MASK>
-                <CLOS_MASK>0xfff</CLOS_MASK>
+                <CLOS_MASK>0x00ff</CLOS_MASK>
+                <CLOS_MASK>0xff00</CLOS_MASK>
+                <CLOS_MASK>0x000f</CLOS_MASK>
+                <CLOS_MASK>0x0001</CLOS_MASK>
+  
             </RDT>
             <HYPERV_ENABLED>y</HYPERV_ENABLED>
             <IOMMU_ENFORCE_SNP>n</IOMMU_ENFORCE_SNP>
@@ -85,7 +74,7 @@
         </cpu_affinity>
         <clos>
             <vcpu_clos>0</vcpu_clos>
-            <vcpu_clos>0</vcpu_clos>
+            <vcpu_clos>1</vcpu_clos>
         </clos>
         <epc_section>
             <base>0</base>
diff --git a/misc/config_tools/data/ehl-crb-b/industry.xml b/misc/config_tools/data/ehl-crb-b/industry.xml
index dbd83e9f7..b215f7d99 100644
--- a/misc/config_tools/data/ehl-crb-b/industry.xml
+++ b/misc/config_tools/data/ehl-crb-b/industry.xml
@@ -18,23 +18,11 @@
             <RDT>
                 <RDT_ENABLED>y</RDT_ENABLED>
                 <CDP_ENABLED>n</CDP_ENABLED>
-                <CLOS_MASK>0xfff</CLOS_MASK>
-                <CLOS_MASK>0xfff</CLOS_MASK>
-                <CLOS_MASK>0xfff</CLOS_MASK>
-                <CLOS_MASK>0xfff</CLOS_MASK>
-                <CLOS_MASK>0xfff</CLOS_MASK>
-                <CLOS_MASK>0xfff</CLOS_MASK>
-                <CLOS_MASK>0xfff</CLOS_MASK>
-                <CLOS_MASK>0xfff</CLOS_MASK>
-                <CLOS_MASK>0xfff</CLOS_MASK>
-                <CLOS_MASK>0xfff</CLOS_MASK>
-                <CLOS_MASK>0xfff</CLOS_MASK>
-                <CLOS_MASK>0xfff</CLOS_MASK>
-                <CLOS_MASK>0xfff</CLOS_MASK>
-                <CLOS_MASK>0xfff</CLOS_MASK>
-                <CLOS_MASK>0xfff</CLOS_MASK>
-                <CLOS_MASK>0xfff</CLOS_MASK>
-            </RDT>
+                <CLOS_MASK>0x00ff</CLOS_MASK>
+                <CLOS_MASK>0xff00</CLOS_MASK>
+                <CLOS_MASK>0x000f</CLOS_MASK>
+                <CLOS_MASK>0x0001</CLOS_MASK>
+             </RDT>
             <HYPERV_ENABLED>y</HYPERV_ENABLED>
             <IOMMU_ENFORCE_SNP>n</IOMMU_ENFORCE_SNP>
             <ACPI_PARSE_ENABLED>y</ACPI_PARSE_ENABLED>
@@ -171,7 +159,7 @@
         </cpu_affinity>
         <clos>
             <vcpu_clos>0</vcpu_clos>
-            <vcpu_clos>0</vcpu_clos>
+            <vcpu_clos>1</vcpu_clos>
         </clos>
         <epc_section>
             <base>0</base>
diff --git a/misc/config_tools/data/tgl-rvp/hybrid_rt.xml b/misc/config_tools/data/tgl-rvp/hybrid_rt.xml
index 628cf35c0..483ec286d 100644
--- a/misc/config_tools/data/tgl-rvp/hybrid_rt.xml
+++ b/misc/config_tools/data/tgl-rvp/hybrid_rt.xml
@@ -17,8 +17,12 @@
         <MULTIBOOT2>y</MULTIBOOT2>
         <ENFORCE_TURNOFF_AC>y</ENFORCE_TURNOFF_AC>
         <RDT>
-            <RDT_ENABLED>n</RDT_ENABLED>
+            <RDT_ENABLED>y</RDT_ENABLED>
             <CDP_ENABLED>n</CDP_ENABLED>
+                <CLOS_MASK>0xff0</CLOS_MASK>
+                <CLOS_MASK>0x00f</CLOS_MASK>
+                <CLOS_MASK>0xff0</CLOS_MASK>
+                <CLOS_MASK>0xff0</CLOS_MASK>
         </RDT>
         <HYPERV_ENABLED>y</HYPERV_ENABLED>
         <IOMMU_ENFORCE_SNP>n</IOMMU_ENFORCE_SNP>
@@ -73,7 +77,7 @@
     </cpu_affinity>
     <clos>
         <vcpu_clos>0</vcpu_clos>
-        <vcpu_clos>0</vcpu_clos>
+        <vcpu_clos>1</vcpu_clos>
     </clos>
     <epc_section>
         <base>0</base>
diff --git a/misc/config_tools/data/tgl-rvp/industry.xml b/misc/config_tools/data/tgl-rvp/industry.xml
index 926be0a7e..d1b64f5af 100644
--- a/misc/config_tools/data/tgl-rvp/industry.xml
+++ b/misc/config_tools/data/tgl-rvp/industry.xml
@@ -17,8 +17,12 @@
         <MULTIBOOT2>y</MULTIBOOT2>
         <ENFORCE_TURNOFF_AC>y</ENFORCE_TURNOFF_AC>
         <RDT>
-            <RDT_ENABLED>n</RDT_ENABLED>
+            <RDT_ENABLED>y</RDT_ENABLED>
             <CDP_ENABLED>n</CDP_ENABLED>
+                <CLOS_MASK>0xff0</CLOS_MASK>
+                <CLOS_MASK>0x00f</CLOS_MASK>
+                <CLOS_MASK>0xff0</CLOS_MASK>
+                <CLOS_MASK>0xff0</CLOS_MASK>
         </RDT>
         <HYPERV_ENABLED>y</HYPERV_ENABLED>
         <IOMMU_ENFORCE_SNP>n</IOMMU_ENFORCE_SNP>
@@ -159,7 +163,7 @@
     </cpu_affinity>
     <clos>
         <vcpu_clos>0</vcpu_clos>
-        <vcpu_clos>0</vcpu_clos>
+        <vcpu_clos>1</vcpu_clos>
     </clos>
     <epc_section>
         <base>0</base>
diff --git a/misc/config_tools/data/tgl-rvp/tgl-rvp.xml b/misc/config_tools/data/tgl-rvp/tgl-rvp.xml
index 2eb667e01..8f7665685 100644
--- a/misc/config_tools/data/tgl-rvp/tgl-rvp.xml
+++ b/misc/config_tools/data/tgl-rvp/tgl-rvp.xml
@@ -280,9 +280,9 @@
 	</TPM_INFO>
 
 	<CLOS_INFO>
-	rdt resources supported: L2
-	rdt resource clos max: 8
-	rdt resource mask max: '0xfffff'
+	rdt resources supported: L3
+	rdt resource clos max: 4
+	rdt resource mask max: '0xfff'
 	</CLOS_INFO>
 
 	<IOMEM_INFO>
diff --git a/misc/config_tools/data/whl-ipc-i7/hybrid_rt.xml b/misc/config_tools/data/whl-ipc-i7/hybrid_rt.xml
index 2ffccd75b..241b6e66b 100644
--- a/misc/config_tools/data/whl-ipc-i7/hybrid_rt.xml
+++ b/misc/config_tools/data/whl-ipc-i7/hybrid_rt.xml
@@ -17,8 +17,12 @@
         <MULTIBOOT2>y</MULTIBOOT2>
         <ENFORCE_TURNOFF_AC>y</ENFORCE_TURNOFF_AC>
         <RDT>
-            <RDT_ENABLED>n</RDT_ENABLED>
+            <RDT_ENABLED>y</RDT_ENABLED>
             <CDP_ENABLED>n</CDP_ENABLED>
+                <CLOS_MASK>0xff00</CLOS_MASK>
+                <CLOS_MASK>0x00ff</CLOS_MASK>
+                <CLOS_MASK>0x00fe</CLOS_MASK>
+                <CLOS_MASK>0x0001</CLOS_MASK>
         </RDT>
         <HYPERV_ENABLED>y</HYPERV_ENABLED>
         <IOMMU_ENFORCE_SNP>n</IOMMU_ENFORCE_SNP>
@@ -70,7 +74,7 @@
     </cpu_affinity>
     <clos>
         <vcpu_clos>0</vcpu_clos>
-        <vcpu_clos>0</vcpu_clos>
+        <vcpu_clos>1</vcpu_clos>
     </clos>
     <epc_section>
         <base>0</base>
diff --git a/misc/config_tools/data/whl-ipc-i7/industry.xml b/misc/config_tools/data/whl-ipc-i7/industry.xml
index c2d4d6e32..2a5dd463b 100644
--- a/misc/config_tools/data/whl-ipc-i7/industry.xml
+++ b/misc/config_tools/data/whl-ipc-i7/industry.xml
@@ -17,8 +17,12 @@
         <MULTIBOOT2>y</MULTIBOOT2>
         <ENFORCE_TURNOFF_AC>y</ENFORCE_TURNOFF_AC>
         <RDT>
-            <RDT_ENABLED>n</RDT_ENABLED>
+            <RDT_ENABLED>y</RDT_ENABLED>
             <CDP_ENABLED>n</CDP_ENABLED>
+                <CLOS_MASK>0xff00</CLOS_MASK>
+                <CLOS_MASK>0x00ff</CLOS_MASK>
+                <CLOS_MASK>0x00fe</CLOS_MASK>
+                <CLOS_MASK>0x0001</CLOS_MASK>
         </RDT>
         <HYPERV_ENABLED>y</HYPERV_ENABLED>
         <IOMMU_ENFORCE_SNP>n</IOMMU_ENFORCE_SNP>
@@ -157,7 +161,7 @@
     </cpu_affinity>
     <clos>
         <vcpu_clos>0</vcpu_clos>
-        <vcpu_clos>0</vcpu_clos>
+        <vcpu_clos>1</vcpu_clos>
     </clos>
     <epc_section>
         <base>0</base>
diff --git a/misc/config_tools/data/whl-ipc-i7/whl-ipc-i7.xml b/misc/config_tools/data/whl-ipc-i7/whl-ipc-i7.xml
index a488a01e2..963e7ae11 100644
--- a/misc/config_tools/data/whl-ipc-i7/whl-ipc-i7.xml
+++ b/misc/config_tools/data/whl-ipc-i7/whl-ipc-i7.xml
@@ -183,6 +183,9 @@
 	</MMCFG_BASE_INFO>
 
 	<CLOS_INFO>
+	rdt resources supported: L3
+	rdt resource clos max: 4
+	rdt resource mask max: '0xffff'
 	</CLOS_INFO>
 
 	<IOMEM_INFO>
-- 
2.17.1

